### ISSUE TITLE
Prefix all CSS styles to prevent style name conflicts

### ISSUE THREAD

**Issue by taurheim** [microsoft/fluentui#11128]
#### Describe the feature that you would like added

Currently merge-styles generates styles with names like `.css-0`. This works very well if all the code on the page comes from one app, but in cases where (for example) a browser extension injects a component that uses fabric and the host page also uses fabric, the style names could conflict and cause unexpected and unpredictable behaviour.

I would like to see a way to prefix all styles from an app, so instead of using `.css-0`, a browser extension could prefix all css names with something like `.ext-css-0`. This way styles would never conflict between two versions/instances of fabric/merge-styles.

#### What component or utility would this be added to

Merge-styles? Maybe a webpack configuration option or a global prefix that can be set?

#### Have you discussed this feature with our team, and if so, who

N/A

#### Additional context/screenshots

It looks like merge-styles already supports something similar "displayName" documented [here](https://www.npmjs.com/package/@uifabric/merge-styles) so maybe the best way would be to expose this on a global level somehow?

**Comment by dzearing** [comment]
Does this work for you?

```
import { Stylesheet } from '@uifabric/merge-styles';

Stylesheet.getInstance().setConfig({ namespace: 'foo' });
```

Example:

https://codesandbox.io/s/affectionate-http-gcrm4

**Comment by taurheim** [comment]
That seems to work for most elements, but for some function calls (e.g. [disableBodyScroll](https://github.com/OfficeDev/office-ui-fabric-react/blob/09d5abf9a89b1f2ae31de5427e24279537eb9a4c/packages/utilities/src/scroll.ts)) they still end up creating styles like `css-0`.

https://codesandbox.io/s/naughty-shockley-5wm3q

**Comment by dzearing** [comment]
Yep, it's a timing issue, which hopefully you can control through script loading. There is an alternative way to do this that can beat timing issues guaranteed, which is to burn a config on the page.

```
<html>
<script type="text/javascript>
window.FabricConfig = {
  mergeStyles: {
    namespace: 'foo'
  }
};
</script>
```

**Comment by taurheim** [comment]
Interesting, in that case it would prefix all instances of fabric in the page, right? (e.g. in the example with a browser extension, the extension would also use that namespace)

**Comment by dzearing** [comment]
I see; so actually, it's less of a global setting you want, and more of a scoped one?

For example, if we provided settings through a React context based provider, would that solve your concern better?

```
<ThemeProvider config={{ namespace: 'foo' }}>
  <YourStuff/>
</ThemeProvider>
```

**Comment by dzearing** [comment]
In the original comments above, you mentioned:

> but in cases where (for example) a browser extension injects a component that uses fabric and the host page also uses fabric, the style names could conflict and cause unexpected and unpredictable behaviour.

I'm not 100% sure exactly which case you're referring to. In merge-styles, the class names are generated by using a global stylesheet manager object attached to the window. If two decoupled pieces of code use 2 different copies of merge-styles, they still share the same manager state, causing the rules to share the same counter id. That is, new rules registered will not conflict with each other.

There are edge cases where we've seen problems. For example when devs don't assume the rules will be unique and create global css rules targeting `css-123` thinking it will never change, obviously that can break very easily and cause conflicts. But that's a bug on the user's end.

**Comment by taurheim** [comment]
Yes, a scoped one was what I was looking for - the react context provider looks like a great solution to this :)

I wasn't aware of the fact that merge-styles used a global stylesheet manager - I saw this interaction between an extension I had installed and my web app and assumed that was what was happening. I wonder if a situation like server-side rendering could be the cause? My web app uses fabric normally but it's possible that the elements & styles that the extension injects were pre-rendered.
